server/functions/framework/compiled_function.go

```diff
@@ -193,232 +201,334 @@ func (c *CompiledFunction) Eval(ctx *sql.Context, row sql.Row) (interface{}, err
-		for paramIdx, matchParam := range match {
-			if parameters[paramIdx].BaseID() != matchParam && matchParam.GetTypeCategory().IsPreferredType(matchParam) {
+		for argIdx := range argTypes {
+			// For variadic functions, the final param consumes all remaining arguments
+			var paramType pgtypes.DoltgresTypeBaseID
+			if argIdx >= len(cand.params.paramTypes) {
+				paramType = cand.params.paramTypes[len(cand.params.paramTypes)-1]
+			} else {
+				paramType = cand.params.paramTypes[argIdx]
+			}
+
+			if argTypes[argIdx].BaseID() != paramType && paramType.GetTypeCategory().IsPreferredType(paramType) {
 				currentPreferredCount++
 			}
 		}
+
 		if currentPreferredCount > preferredCount {
 			preferredCount = currentPreferredCount
-			preferredOverloads = append([][]pgtypes.DoltgresTypeBaseID{}, match)
-			preferredCasts = append([][]TypeCastFunction{}, matchCasts[matchIdx])
+			preferredOverloads = append([]overloadMatch{}, cand)
 		} else if currentPreferredCount == preferredCount {
-			preferredOverloads = append(preferredOverloads, match)
-			preferredCasts = append(preferredCasts, matchCasts[matchIdx])
+			preferredOverloads = append(preferredOverloads, cand)
 		}
 	}
-	// Check once more for exactly one match
-	if len(preferredOverloads) == 1 {
-		matchedOverload := c.Functions
-		for _, parameter := range preferredOverloads[0] {
-			matchedOverload = matchedOverload.Parameter[parameter]
+	return preferredOverloads
+}
+
+// exactTypeMatches returns the set of overload candidates that have the most exact type matches for the arg types
+// provided.
+func exactTypeMatches(argTypes []pgtypes.DoltgresType, candidates []overloadMatch) []overloadMatch {
+	matchCount := 0
+	var matches []overloadMatch
+	for _, cand := range candidates {
+		currentMatchCount := 0
+		for argIdx := range argTypes {
+			// For variadic functions, the final param consumes all remaining arguments
+			var paramType pgtypes.DoltgresTypeBaseID
+			if argIdx >= len(cand.params.paramTypes) {
+				paramType = cand.params.paramTypes[len(cand.params.paramTypes)-1]
+			} else {
+				paramType = cand.params.paramTypes[argIdx]
+			}
+
+			// NULL values count as exact matches, since all types accept NULL as a valid value
+			paramBaseID := argTypes[argIdx].BaseID()
+			if paramBaseID == paramType || paramBaseID == pgtypes.DoltgresTypeBaseID_Null {
+				currentMatchCount++
+			}
+		}
+		if currentMatchCount > matchCount {
+			matchCount = currentMatchCount
+			matches = append([]overloadMatch{}, cand)
+		} else if currentMatchCount == matchCount {
+			matches = append(matches, cand)
+		}
+	}
+	return matches
+}
+
+// typeCompatibleOverloads returns all overloads that have a matching number of params whose types can be
+// implicitly converted to the ones provided. This is the set of all possible overloads that could be used with the
+// param types provided.
+func (c *CompiledFunction) typeCompatibleOverloads(argTypes []pgtypes.DoltgresType, sources []Source) []overloadMatch {
```

Comment:
I'd rename this to something like `matchingLengthOverloads`, since that goal takes higher precedence than anything else in this subfunction.

Reply:
I changed the method that generates overloads for a given param number to only return overloads that are compatible with that number of arguments, so the length is now implicit. See what you think.